packages/flutter_tools/lib/src/test/coverage_collector.dart

```diff
@@ -129,11 +129,16 @@ class CoverageCollector extends TestWatcher {
 
     final Stopwatch? collectTestTimeRecorderStopwatch = testTimeRecorder?.start(TestTimePhases.CoverageCollect);
 
-    final Future<void> processComplete = testDevice.finished.catchError(
-      (Object error) => throw Exception(
-          'Failed to collect coverage, test device terminated prematurely with '
-          'error: ${(error as TestDeviceException).message}.'),
-      test: (Object error) => error is TestDeviceException,
+    final Future<void> processComplete = testDevice.finished.then(
+      (Object? obj) => obj,
+      onError: (Object error, StackTrace stackTrace) {
+        if (error is TestDeviceException) {
+          throw Exception(
+            'Failed to collect coverage, test device terminated prematurely with '
+            'error: ${error.message}.');
```

Comment:
Consider retaining `stackTrace`.

Reply:
Updated here, and throughout to ensure we're surfacing stacktraces.